File: global-hooks/framework/automation/subagent_tracker.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.8"
# dependencies = []
# ///
"""
Subagent Tracker - SubagentStop Hook

Lightweight tracker that logs subagent execution metadata (agent name,
duration, output length) to ~/.claude/data/agent_tracking.jsonl for
observability and quality analysis.

Hook event: SubagentStop
Exit codes:
  0 = Always (never blocks, tracking only)
"""

import json
import re
import sys
from datetime import datetime, timezone
from pathlib import Path

# Output below this character count is considered suspiciously empty
MINIMUM_MEANINGFUL_OUTPUT = 50

# Patterns that signal a failed/errored agent response
ERROR_PATTERNS = [
    "traceback", "exception:", "error:", "internal error",
    "rate_limit", "authentication_failed", "tool_call_error",
    "context_length_exceeded", "max_tokens", "500 ", " 500\n",
    "i cannot", "i'm unable", "i am unable",
]

# CAF role agent name prefixes — these have expected output files we can verify
CAF_ROLE_AGENTS = {"builder", "validator", "debugger"}


def get_tracking_path():
    """Return path to agent tracking JSONL file, creating dirs if needed."""
    tracking_path = Path.home() / ".claude" / "data" / "agent_tracking.jsonl"
    tracking_path.parent.mkdir(parents=True, exist_ok=True)
    return tracking_path


def main():
    try:
        raw = sys.stdin.read()
        if not raw.strip():
            sys.exit(0)

        hook_input = json.loads(raw)

        # Extract subagent metadata from hook payload
        tool_name = hook_input.get("tool_name", "unknown")
        tool_input = hook_input.get("tool_input", {})
        tool_output = hook_input.get("tool_output", "")

        # Compute output length
        if isinstance(tool_output, str):
            output_length = len(tool_output)
        elif isinstance(tool_output, dict):
            output_length = len(json.dumps(tool_output))
        else:
            output_length = 0

        # Extract agent name from tool input if available
        agent_name = "unknown"
        if isinstance(tool_input, dict):
            agent_name = (
                tool_input.get("agent_name")
                or tool_input.get("name")
                or tool_input.get("task_description", "unknown")[:80]
            )
        elif isinstance(tool_input, str):
            agent_name = tool_input[:80]

        # ── Anomaly Detection ─────────────────────────────────────────────
        output_text = tool_output if isinstance(tool_output, str) else json.dumps(tool_output)

        is_empty_output = output_length < MINIMUM_MEANINGFUL_OUTPUT

        has_error_signal = any(p in output_text.lower() for p in ERROR_PATTERNS)

        # Check if this is a CAF role agent with an expected output file
        output_file_missing = False
        expected_file = None
        match = re.search(r'(builder|validator|debugger)-(\d+)', agent_name.lower())
        if match:
            role, iteration = match.group(1), match.group(2)
            expected_file = f"/tmp/caf_{role}_{iteration}.md"
            output_file_missing = not Path(expected_file).exists()

        anomaly_types = []
        if is_empty_output:
            anomaly_types.append("empty_output")
        if has_error_signal:
            anomaly_types.append("error_signal")
        if output_file_missing:
            anomaly_types.append("missing_output_file")

        # ── Build tracking record ─────────────────────────────────────────
        record = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "agent_name": agent_name,
            "tool_name": tool_name,
            "output_length": output_length,
            "input_keys": list(tool_input.keys()) if isinstance(tool_input, dict) else [],
            "anomalies": anomaly_types,
        }

        # Append to tracking file
        tracking_path = get_tracking_path()
        with open(tracking_path, "a") as f:
            f.write(json.dumps(record) + "\n")

        # ── Surface anomalies to watchdog and alert log ───────────────────
        if anomaly_types:
            ts = record["timestamp"]

            # Write to watchdog state file so agent-watchdog detects it
            # without requiring the failing agent to self-report
            watchdog_path = Path("/tmp/caf_watchdog.md")
            error_detail = "+".join(anomaly_types)
            if has_error_signal:
                error_detail += f":{output_text[:120].replace(chr(10), ' ')}"
            watchdog_line = (
                f"[{ts}] AGENT:{agent_name} STATUS:FAILED "
                f"TASK:hook_detected ERROR:{error_detail}\n"
            )
            try:
                with open(watchdog_path, "a") as f:
                    f.write(watchdog_line)
            except Exception:
                pass  # Never fail the hook

            # Write to persistent alert log for post-session review
            alert_path = Path.home() / ".claude" / "data" / "subagent_alerts.jsonl"
            alert = {
                "timestamp": ts,
                "agent_name": agent_name,
                "anomaly_types": anomaly_types,
                "output_length": output_length,
                "expected_file": expected_file,
                "error_excerpt": output_text[:200] if has_error_signal else None,
            }
            try:
                with open(alert_path, "a") as f:
                    f.write(json.dumps(alert) + "\n")
            except Exception:
                pass  # Never fail the hook

    except Exception:
        # Never fail, never block
        pass

    sys.exit(0)


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass

File: caf-hooks/src/hooks/subagent_tracker.rs
/// Subagent Tracker — SubagentStop hook.
///
/// Python equivalent: global-hooks/framework/automation/subagent_tracker.py
///
/// Behavior:
/// - Reads SubagentStop event from stdin JSON
/// - Extracts agent name, tool name, and output from the payload
/// - Checks output for error/anomaly patterns
/// - Appends tracking record to ~/.claude/data/agent_tracking.jsonl
/// - On anomalies: appends to /tmp/caf_watchdog.md and ~/.claude/data/subagent_alerts.jsonl
/// - Always exits 0 (never blocks)
use std::path::PathBuf;

use chrono::Utc;
use regex::Regex;
use serde_json::Value;

use crate::io::{read_stdin_value, try_append_jsonl, write_output};
use crate::types::HookOutput;

/// Output below this character count is considered suspiciously empty
const MINIMUM_MEANINGFUL_OUTPUT: usize = 50;

/// Patterns that signal a failed/errored agent response (match Python exactly)
const ERROR_PATTERNS: &[&str] = &[
    "traceback",
    "exception:",
    "error:",
    "internal error",
    "rate_limit",
    "authentication_failed",
    "tool_call_error",
    "context_length_exceeded",
    "max_tokens",
    "500 ",
    " 500\n",
    "i cannot",
    "i'm unable",
    "i am unable",
];

fn tracking_path() -> PathBuf {
    let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));
    home.join(".claude").join("data").join("agent_tracking.jsonl")
}

fn alerts_path() -> PathBuf {
    let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));
    home.join(".claude").join("data").join("subagent_alerts.jsonl")
}

pub fn run() {
    let data: Value = read_stdin_value();

    // Extract fields matching Python logic
    let tool_name = data
        .get("tool_name")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();

    let tool_input = data.get("tool_input").cloned().unwrap_or(Value::Null);
    let tool_output = data.get("tool_output").cloned().unwrap_or(Value::Null);

    // Compute output length (matches Python: str→len, dict→json len, else 0)
    let output_text = match &tool_output {
        Value::String(s) => s.clone(),
        Value::Null => String::new(),
        other => serde_json::to_string(other).unwrap_or_default(),
    };
    let output_length = output_text.len();

    // Extract agent name from tool input (matches Python priority order)
    let agent_name = match &tool_input {
        Value::Object(map) => {
            map.get("agent_name")
                .or_else(|| map.get("name"))
                .and_then(|v| v.as_str())
                .map(|s| s.to_string())
                .or_else(|| {
                    map.get("task_description")
                        .and_then(|v| v.as_str())
                        .map(|s| s.chars().take(80).collect())
                })
                .unwrap_or_else(|| "unknown".to_string())
        }
        Value::String(s) => s.chars().take(80).collect(),
        _ => "unknown".to_string(),
    };

    // Input keys for the record
    let input_keys: Vec<String> = match &tool_input {
        Value::Object(map) => map.keys().cloned().collect(),
        _ => Vec::new(),
    };

    // ── Anomaly Detection ──────────────────────────────────────────────────
    let is_empty_output = output_length < MINIMUM_MEANINGFUL_OUTPUT;

    let output_lower = output_text.to_lowercase();
    let has_error_signal = ERROR_PATTERNS.iter().any(|p| output_lower.contains(p));

    // Check if this is a CAF role agent with an expected output file
    let re = Regex::new(r"(builder|validator|debugger)-(\d+)").unwrap();
    let (output_file_missing, expected_file) =
        if let Some(caps) = re.captures(&agent_name.to_lowercase()) {
            let role = caps.get(1).map(|m| m.as_str()).unwrap_or("");
            let iteration = caps.get(2).map(|m| m.as_str()).unwrap_or("");
            let path = format!("/tmp/caf_{}_{}.md", role, iteration);
            let missing = !std::path::Path::new(&path).exists();
            (missing, Some(path))
        } else {
            (false, None)
        };

    let mut anomaly_types: Vec<String> = Vec::new();
    if is_empty_output {
        anomaly_types.push("empty_output".to_string());
    }
    if has_error_signal {
        anomaly_types.push("error_signal".to_string());
    }
    if output_file_missing {
        anomaly_types.push("missing_output_file".to_string());
    }

    // ── Build tracking record ──────────────────────────────────────────────
    let timestamp = Utc::now().to_rfc3339();
    let record = serde_json::json!({
        "timestamp": timestamp,
        "agent_name": agent_name,
        "tool_name": tool_name,
        "output_length": output_length,
        "input_keys": input_keys,
        "anomalies": anomaly_types,
    });

    let _ = try_append_jsonl(&tracking_path(), &record);

    // ── Surface anomalies ──────────────────────────────────────────────────
    if !anomaly_types.is_empty() {
        // Write to watchdog state file
        let watchdog_path = std::path::Path::new("/tmp/caf_watchdog.md");
        let mut error_detail = anomaly_types.join("+");
        if has_error_signal {
            let excerpt: String = output_text
                .chars()
                .take(120)
                .collect::<String>()
                .replace('\n', " ");
            error_detail.push(':');
            error_detail.push_str(&excerpt);
        }
        let watchdog_line = format!(
            "[{}] AGENT:{} STATUS:FAILED TASK:hook_detected ERROR:{}\n",
            timestamp, agent_name, error_detail
        );
        let _ = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(watchdog_path)
            .and_then(|mut f| {
                use std::io::Write;
                f.write_all(watchdog_line.as_bytes())
            });

        // Write to persistent alert log
        let error_excerpt = if has_error_signal {
            Some(output_text.chars().take(200).collect::<String>())
        } else {
            None
        };
        let alert = serde_json::json!({
            "timestamp": timestamp,
            "agent_name": agent_name,
            "anomaly_types": anomaly_types,
            "output_length": output_length,
            "expected_file": expected_file,
            "error_excerpt": error_excerpt,
        });
        let _ = try_append_jsonl(&alerts_path(), &alert);
    }

    write_output(&HookOutput::empty());
}
